module/zfs/vdev_raidz.c

```diff
@@ -2395,11 +3404,48 @@ vdev_raidz_io_done(zio_t *zio)
 {
 	raidz_map_t *rm = zio->io_vsd;
 
+	ASSERT(zio->io_bp != NULL);
 	if (zio->io_type == ZIO_TYPE_WRITE) {
 		for (int i = 0; i < rm->rm_nrows; i++) {
 			vdev_raidz_io_done_write_impl(zio, rm->rm_row[i]);
 		}
 	} else {
+		if (rm->rm_phys_col) {
+			/*
+			 * This is an aggregated read.  Copy the data and status
+			 * from the aggregate abd's to the individual rows.
+			 */
+			for (int i = 0; i < rm->rm_nrows; i++) {
+				raidz_row_t *rr = rm->rm_row[i];
+
+				for (int c = 0; c < rr->rr_cols; c++) {
+					raidz_col_t *rc = &rr->rr_col[c];
+					if (rc->rc_size == 0)
```

Comment:
```suggestion
					if (rc->rc_tried || rc->rc_size == 0)
```

We want to be careful to not always copy the aggregated read `->rm_phys_col` columns and `rc_*` flags to the normal row columns.  Otherwise it's possible to get caught in an infinite loop where we incorrectly clear the `rc->rc_tried` flag for any column which was read by the `vdev_raidz_read_all()` in the `zio_vdev_io_redone()` return path.  We'll also overwrite the column with bad data from the original aggregated read.  We should skip the column if `rc->rc_tried` is set.  I managed to trigger this case when doing some manual testing.

Reply:
Great find!  Thanks for chasing this one down. Fixed